ethers/providers/jsonrpc.nim

```diff
@@ -165,6 +189,91 @@ method signMessage*(signer: JsonRpcSigner,
   return await client.eth_sign(address, message)
 
 method sendTransaction*(signer: JsonRpcSigner,
-                        transaction: Transaction) {.async.} =
-  let client = await signer.provider.client
-  discard await client.eth_sendTransaction(transaction)
+                        transaction: Transaction): Future[TransactionResponse] {.async.} =
+  let
+    client = await signer.provider.client
+    hash = await client.eth_sendTransaction(transaction)
+
+  return TransactionResponse(hash: hash, provider: signer.provider)
+
+method wait*(tx: TransactionResponse,
+             wantedConfirms = DEFAULT_CONFIRMATIONS,
+             timeoutInBlocks = RECEIPT_TIMEOUT_BLKS.some): # will error if tx not mined in x blocks
+            Future[TransactionReceipt]
+            {.async, upraises: [JsonRpcProviderError].} = # raises for clarity
+
+  var
+    receipt: ?TransactionReceipt
+    subscription: JsonRpcSubscription
+
+  let
+    provider = JsonRpcProvider(tx.provider)
+    retFut = newFuture[TransactionReceipt]("wait")
+
+  proc confirmations(receipt: TransactionReceipt, atBlkNum: UInt256): UInt256 =
+
+    var confirms = (atBlkNum - !receipt.blockNumber) + 1
```

Comment:
The `!` was for `receipt`, which we know was always present as there was a loop that waited for it to be not `.isNone`. Also, `Block.number` is a `UInt256` and according to `conversion.nim`, this would be marshalled as:
```nim
UInt256.fromHex(json.getStr())
```
which would always be `0` if the node did not emit a block number in the `newHead` event.

I've refactored the closure proc a bit to make it more clear, and I'll add a test the emulates the node not returning a block number in the `newHead` event.

Reply:
Also, the while loop checked that receipt.blockNumber existed as well. Again, not super clear, so hopefully the refactor clears it up.